### DISCUSSION TITLE
How to ask for access permissions with an OSx App?

### DISCUSSION THREAD

**Discussion by hnuecke** [airsdk/Adobe-Runtime-Support#1223] [Q&A]
Since Catalina access to the _documents folder_ is no longer granted.
The user needs to explicitely accept, that the app geta access rights (permission) to read/write data under the documents folder.

Native Apps obviously pop-up a window, asking the user to grant such rights.

**My question is:**
How can I achieve this with an AIR app?

Is there a setting in the config file or needs that to be set in the packaging tool?

Thanks for any guidance!!

**Comment by SUSDEV** [comment]
Doesn't your app make the request when it tries to read/access Documents? I have an app with at least a few hundred installs in Catalina and beyond, and I don't think anyone has called me with Documents not being accessible. I do get calls occasionally from iOS users who don't have access to the CameraRoll and so I have to have them set it manually, but there is a CameraRoll object with a permissionStatus member that I check.

**Reply by hnuecke** [reply]
Sorry for a probably dumb question ;-)
But: What do you mean by "doesn't your app make a request"?

I use(d) a dedicated folder with the name of my app in the documents folder, where I managed all data for the user of my app.
I.e. if that specific "App-folder" does not exist, it is created programmatically.
At start-up, if that folder exists, the content is read: fileArray = appFolder.getDirectoryListing()
And that statement now fails silently: A "try-catch" passes without error, but the array is empty. Despite the fact that there ARE files and folders available.

I now use a folder directly at the user folder, which works fine and is an acceptable "workaround".

Nevertheless, I'd like to know how to ask the user granting access to the documents folder for the running app; triggered by/under control of the App.

I hope this makes sense, and you understand my  "problem" better now...

Thanks and regards

**Reply by SUSDEV** [reply]
I understand, not dumb question at all. I last compiled and released my Mac app in Sep 2020 on Catalina. I don't know if something changed in the meantime, but I have a similar situation to you except I look for a specific file to load (which points to other files) rather than getDirectoryListing(). This sounds like an @ajwfrost question to look into/address. Unfortunately I don't think I will be updating my Mac app for at least a month or two so I can't see if I get the same behavior as you with a later AIR build.

**Reply by ajwfrost** [reply]
Hi @hnuecke  - this functionality should be automatic within Catalina/later, as long as the application is signed properly?

You can add an (optional?) message to the user which is added to the dialog that's then displayed, via the app descriptor file:
```
        <macOS>
                <InfoAdditions><![CDATA[
                        <key>NSDocumentsFolderUsageDescription</key>
                        <string>We need to access the documents folder</string>
                ]]></InfoAdditions>
        </macOS>
```

But if you codesign the generated app/bundle, it should then prompt the user for the permission.

Signing can be done by adding 'native' signing information on the command line, or later on by using the `adt -sign` method. E.g.
```
adt -package [signing options] -target bundle -storetype KeychainStore -alias "Developer ID Application" OutputBundleName application.xml [files to package..]
```

thanks

**Reply by hnuecke** [reply]
@ajwfrost  Thank you, very valuable information! BTW, how can I mark your reply as "answer"?
@SUSDEV  And thanks for getting ajwfrost in the loop ;-)

When I find the time (after I "solved" the issue "pragmatically", my first priority now is selling my App ;-) ) I'll have a closer look and will  add the lines to the app descriptor file to see what happens!
I already sign my app when building the package. I have a valid code sign certificate., So I guess that step should be OK?!

I do not code sign (again) when creating the final package. I use "Packages" (http://s.sudre.free.fr/Software/Packages/about.html) to build the .pkg file and do not use Apple's shop.

If anybody has a suggestion for an alternative Packager that support code signing the final package... I would be more than happy!

I do feel more comfortable in the Windows world, and know how to sign the final installer (.exe, .msi or .msix) there, Buti n the MAC world I'm a stranger...
